volmaster/daemon.go

```diff
@@ -423,30 +423,38 @@ func (d *DaemonConfig) handleRemove(w http.ResponseWriter, r *http.Request) {
 		Reason: lock.ReasonRemove,
 	}
 
+	etcdRemove := func() error {
+		if err := d.Config.RemoveVolume(req.Policy, req.Volume); err != nil {
+			return errored.Errorf("Clearing volume records for %q", vc).Combine(err)
+		}
+
+		return nil
+	}
+
+	complete := func() error {
+		if err := d.removeVolume(vc, d.Global.Timeout); err != nil && err != errNoActionTaken {
+			log.Warn(errored.Errorf("Removing image %q", vc).Combine(err))
+		}
+
+		return etcdRemove()
+	}
+
 	err = lock.NewDriver(d.Config).ExecuteWithMultiUseLock([]config.UseLocker{uc, snapUC}, d.Global.Timeout, func(ld *lock.Driver, ucs []config.UseLocker) error {
 		exists, err := d.existsVolume(vc)
 		if err != nil && err != errNoActionTaken {
 			return err
 		}
 
 		if err == errNoActionTaken {
-			goto publish
+			return complete()
 		}
 
 		if !exists {
+			etcdRemove()
```

Comment:
I might be reading this wrong, but may be this should be `return etcdRemove()` ? Otherwise I think we will fall through and call `complete()` below which will call `etcdRemove()` again and fail.